snakemake_executor_plugin_latch/resource_utils.py

```diff
@@ -0,0 +1,216 @@
+import re
+from dataclasses import dataclass
+from typing import Optional
+
+from humanfriendly import format_size, parse_size
+
+_image_expr = re.compile(r"^(([a-zA-Z]+)://)?(?P<uri>[^/]+.*)$")
+
+
+def sanitize_image_name(image: str) -> str:
+    match = _image_expr.match(image)
+    if match is None:
+        raise ValueError(f"malformed image name: {image}")
+
+    return match["uri"]
+
+
+@dataclass
+class Resources:
+    cpu: int = 2000  # millicores
+    mem: int = 4 * 1024**3  # bytes
+    disk: int = 500 * 1024**3  # bytes
+    gpus: int = 0
+    gpu_type: Optional[str] = None
+
+    def __str__(self) -> str:
+        return ", ".join(
+            [
+                f"{self.cpu / 1000:.3g} CPUs",
+                f"{format_size(self.mem)} RAM",
+                f"{format_size(self.disk)} Storage",
+                *(
+                    [f"{self.gpus} {self.gpu_type} GPU(s)"]
+                    if self.gpus > 0 and self.gpu_type is not None
+                    else []
+                ),
+            ]
+        )
+
+    def __eq__(self, other):
```

Comment:
dont dataclasses already implement __eq__ by checking equality of fields?

Reply:
oh duh